mesh/mesh.go

```diff
@@ -281,30 +225,40 @@ func (msh *Mesh) setProcessedLayer(layerID types.LayerID) error {
 	}
 	msh.processedLayer.Store(processed)
 	events.ReportNodeStatusUpdate()
-	msh.logger.Event().Info("processed layer set", processed)
+	logger.Event().Info("processed layer set", processed)
 	return nil
 }
 
-func (msh *Mesh) revertMaybe(ctx context.Context, logger log.Log, newVerified types.LayerID) error {
-	minUpdated := msh.getMinUpdatedLayer()
-	if minUpdated == (types.LayerID{}) {
-		// no contextual validity update since the last ProcessLayer() call
-		return nil
-	}
+func (msh *Mesh) processValidityUpdates(ctx context.Context, logger log.Log, newVerified types.LayerID, updated []types.BlockContextualValidity) error {
+	logger.With().Debug("got validity update",
+		log.Int("num_updates", len(updated)),
+		log.Array("updates", log.ArrayMarshalerFunc(func(encoder log.ArrayEncoder) error {
+			for _, bv := range updated {
+				encoder.AppendString(bv.ID.String())
+				encoder.AppendBool(bv.Validity)
+			}
+			return nil
+		})))
 
-	msh.resetMinUpdatedLayer(minUpdated)
-	if minUpdated.After(msh.LatestLayerInState()) {
-		// nothing to do
-		return nil
+	inState := msh.LatestLayerInState()
+	byLayer := make(map[types.LayerID][]types.BlockContextualValidity)
```

Comment:
it is also possible to maintain a map of updates in the tortoise if you need it instead of list

in general i would try to avoid transformations and prepare a data structure that works for intended use case

Reply:
for the corner case stated here https://github.com/spacemeshos/go-spacemesh/pull/3791#discussion_r1032649571
i ended up keeping a map in mesh.go to remember update failures. so i keep tortoise interface as a slice as it's simpler.